pkg/types/ccipocr3/chain_accessor.go

```diff
@@ -0,0 +1,313 @@
+package ccipocr3
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	rmntypes "github.com/smartcontractkit/chainlink-ccip/commit/merkleroot/rmn/types"
+	iplugintypes "github.com/smartcontractkit/chainlink-ccip/internal/plugintypes"
+	"github.com/smartcontractkit/chainlink-ccip/plugintypes"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+// AccessorMetadata contains metadata about the chain accessor.
+type AccessorMetadata struct {
+	ChainSelector ChainSelector
+	Contracts     map[string]UnknownAddress
+}
+
+// ChainAccessor for all direct chain access. The accessor is responsible for
+// in addition to direct access to the chain, this interface also translates
+// onchain representations of data to the plugin representation.
+type ChainAccessor interface {
+	AllAccessors
+	SourceAccessor
+	DestinationAccessor
+	RMNAccessor
+}
+
+// AllAccessors contains functionality that is available to all types of accessors.
+type AllAccessors interface {
+	// Metadata associated with the chain accessor. Replaces GetContractAddress.
+	Metadata() AccessorMetadata
+
+	// TODO: Discovery/Binding functions here?
+
+	// GetContractAddress returns the contract address that is registered for the provided contract name and chain.
+	// WARNING: This function will fail if the oracle does not support the requested chain.
+	//
+	// Deprecated: use Metadata() instead.
+	GetContractAddress(contractName string) ([]byte, error)
+
+	// GetChainFeeComponents Returns all fee components for given chains if corresponding
+	// chain writer is available.
+	//
+	// Access Type: ChainWriter
+	// Contract: N/A
+	// Confidence: N/A
+	GetChainFeeComponents(
+		ctx context.Context,
+	) map[ChainSelector]ChainFeeComponents
+
+	// GetDestChainFeeComponents seems redundant. If the error is needed lets
+	// add it to GetChainFeeComponents.
+	//
+	// Deprecated: use GetChainFeeComponents instead.
+	GetDestChainFeeComponents(
+		ctx context.Context,
+	) (types.ChainFeeComponents, error)
+
+	// DiscoverContracts reads the destination chain for contract addresses. They are returned per
+	// contract and source chain selector.
+	// allChains is needed because there is no way to enumerate all chain selectors on Solana. We'll attempt to
+	// fetch the source config from the offramp for each of them.
+	DiscoverContracts(ctx context.Context, allChains []ChainSelector) (ContractAddresses, error)
+
+	// Sync can be used to perform frequent syncing operations inside the reader implementation.
+	// Returns a bool indicating whether something was updated.
+	Sync(ctx context.Context, contracts ContractAddresses) error
```

Comment:
This interface seems fine as is. Except `DiscoverContracts` is a ConfigPoller wrapper so maybe it goes away and the interface only has a `Sync` function. Also, might rename `Sync` to `UpdateAddresses`.